### PR TITLE
Add an endpoint for providing the transaction history and overview

### PR DESCRIPTION
**Pull request by MichaelKim20**
I saw the plan of the wallet and realized it to be as supportive as possible.

Relates to #212

### CODE REVIEW

**Comment by MichaelKim20** [comment]
I added two endpoints.

### 1. GET /wallet/transactions/history/:addresses
An endpoint that provides a history of multiple transactions

- The parameter `addresses` are the form of multiple addresses separated by `,`.
- The parameter `pageSize` is the maximum size that can be obtained from one query, default is 10
- The parameter `page` is the number on the page, this value begins with 1, default is 1
- The return type is array of ITxHistoryElement.

```TypeScript
/**
 * The interface of the transactions history element
 */
export interface ITxHistoryElement
{
    /**
     * Address, Public key
     */
    address: string;

    /**
     * Block height
     */
    height: string;

    /**
     * Transaction time
     */
    time: number;

    /**
     * Transaction hash
     */
    tx_hash: string;

    /**
     * Transaction type
     */
    type: number;

    /**
     * Amount
     */
    amount: string;

    /**
     * Block height at which the output of the transaction becomes available
     */
    unlock_height: string;

    /**
     * Time at which the output of the transaction becomes available
     */
    unlock_time: number;
}
```
---
### 2. GET /wallet/transaction/overview/:hash
An endpoint that provides an overview of one transaction
- The parameter `hash` is the hash of the transaction
- The return type is ITxOverview.
``` TypeScript
/**
 * The interface of the transaction overview
 */
export interface ITxOverview
{
    /**
     * Block height
     */
    height: string;

    /**
     * Transaction time
     */
    time: number;

    /**
     * Transaction hash
     */
    tx_hash: string;

    /**
     * Transaction type
     */
    type: number;

    /**
     * Block height at which the output of the transaction becomes available
     */
    unlock_height: string;

    /**
     * Time at which the output of the transaction becomes available
     */
    unlock_time: number;

    /**
     * The address and amount of the output associated with the transaction input
     */
    senders: Array<ITxOverviewElement>;

    /**
     * The address and amount of transaction output
     */
    receivers: Array<ITxOverviewElement>;

    /**
     * Transaction fee
     */
    fee: string;
}

/**
 * The interface of the transaction overview element
 */
export interface ITxOverviewElement
{
    /**
     * Address, Public key
     */
    address: string;

    /**
     * Amount
     */
    amount: string;
}
```

**Comment by MichaelKim20** [comment]
The file below is the database I used to test the query.

[main.db.zip](https://github.com/bpfkorea/stoa/files/5664202/main.db.zip)
[transactions-history5.sql.zip](https://github.com/bpfkorea/stoa/files/5664223/transactions-history5.sql.zip)

File: src/modules/storage/LedgerStorage.ts [hunk 1]
```diff
@@ -849,4 +849,137 @@ export class LedgerStorage extends Storages
...
+                        SUM(S.amount) as spend
+                    FROM
+                        transactions AS T,
+                        tx_inputs AS I,
+                        tx_outputs AS S
+                    WHERE
+                        T.block_height = I.block_height
+                        AND T.tx_index = I.tx_index
+                        AND I.utxo = S.utxo_key
+                        AND S.address in (${accounts})
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]: You must also join the blocks table.
It's not on the block table, but there may only be transactions, tx_input, tx_ouputs

Comment (TrustHenry) [reply src/modules/storage/LedgerStorage.ts]: Is it the standard JOIN method for sqlite?
Shouldn't we do this?
FROM transactions T INNER JOIN tx_inputs I ON (T.block_height = I.block_height)

Comment (MichaelKim20) [reply src/modules/storage/LedgerStorage.ts]: > You must also join the blocks table.
> It's not on the block table, but there may only be transactions, tx_input, tx_ouputs

When is that the case?

Comment (MichaelKim20) [reply src/modules/storage/LedgerStorage.ts]: > Is it the standard JOIN method for sqlite?
> Shouldn't we do this?
> FROM transactions T INNER JOIN tx_inputs I ON (T.block_height = I.block_height)

That's a good idea.

Comment (TrustHenry) [reply src/modules/storage/LedgerStorage.ts]: For data integrity.

File: src/modules/storage/LedgerStorage.ts [hunk 2]
```diff
@@ -849,4 +849,137 @@ export class LedgerStorage extends Storages
...
+            (
+                    SELECT
+                        S.address,
+                        T.block_height,
+                        T.tx_hash,
+                        T.tx_index,
+                        T.type,
+                        T.unlock_height,
+                        0 as income,
+                        SUM(S.amount) as spend
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
                        IFNULL(SUM(S.amount),0) as spend
```

File: src/modules/storage/LedgerStorage.ts (hunk 1 above)
Comment (MichaelKim20) [reply src/modules/storage/LedgerStorage.ts]: Okay, That's a good idea

File: src/modules/storage/LedgerStorage.ts [hunk 3]
```diff
@@ -849,4 +849,137 @@ export class LedgerStorage extends Storages
...
+                    UNION ALL
+
+                    SELECT
+                        O.address,
+                        T.block_height,
+                        T.tx_hash,
+                        T.tx_index,
+                        T.type,
+                        T.unlock_height,
+                        SUM(O.amount) as income,
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
                        IFNULL(SUM(O.amount),0) as income,
```

File: src/modules/storage/LedgerStorage.ts [hunk 4]
```diff
@@ -849,4 +849,137 @@ export class LedgerStorage extends Storages
...
+                        tx_outputs AS O
+                    WHERE
+                        T.block_height = O.block_height
+                        AND T.tx_index = O.tx_index
+                        AND O.address in (${accounts})
+                    GROUP BY T.tx_hash, O.address
+            ) AS TX
+            GROUP BY TX.block_height, TX.tx_hash, TX.address, TX.type
+            ORDER BY TX.block_height DESC, TX.tx_index DESC
+            limit ? offset ?;`;
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
            LIMIT ? OFFSET ?;`;
```

Comment (MichaelKim20) [reply src/modules/storage/LedgerStorage.ts]: @TrustHenry Thank you for review
I changed

**Comment by MichaelKim20** [comment]
I changed the layout of the table for the integrity of the data
And it exists at the beginning of the commit.

File: docs/Database_Schema.md [hunk 5]
```diff
@@ -67,9 +67,10 @@ CREATE TABLE IF NOT EXISTS "enrollments" (
 
 | Column            | Data Type | PK | Not NULL | Default  |Description|
 |:----------------- |:--------- |:--:|:--------:| -------- | --------- |
-| block_height      | INTEGER   | Y  | Y        |          | The height of the block|
-| tx_index          | INTEGER   | Y  | Y        |          | The index of transaction in the block |
-| tx_hash           | BLOB      |    | Y        |          | The hash of transaction |
+| tx_hash           | BLOB      | Y  | Y        |          | The hash of transaction |
+| tx_index          | INTEGER   |    | Y        |          | The index of transaction in the block |
+| block_hash        | BLOB      |    | Y        |          | The hash of the current block |
```
Comment (TrustHenry) [inline docs/Database_Schema.md]: This column is unnecessary from the db normalization point of view.
The primary key of the block is still defined by `height`.
If you want to use this as an FK, you must change the PK in the block table to Block_hash.

Comment (MichaelKim20) [reply docs/Database_Schema.md]: That's a good point.
If I modify this, the changes are too big, so I will create a new issue and proceed.
I will remove the commit from the current PR.

File: src/Stoa.ts [hunk 6]
```diff
@@ -377,6 +379,143 @@ class Stoa extends WebService
...
+     * GET /wallet/transaction/history/:addresses
+     *
+     * Called when a request is received through the `/transaction_history/:addresses` handler
+     * The parameter `addresses` are the form of multiple addresses separated by `,`.
+     * The parameter `pageSize` is the maximum size that can be obtained from one query, default is 10
+     * The parameter `page` is the number on the page, this value begins with 1, default is 1
+     *
+     * Returns a set of transactions history of the addresses.
+     */
+    private getWalletTransactionHistory (req: express.Request, res: express.Response)
```
Comment (TrustHenry) [inline src/Stoa.ts]:
```suggestion
    private getWalletTransactionsHistory (req: express.Request, res: express.Response)
```

File: src/modules/storage/LedgerStorage.ts [hunk 7]
```diff
@@ -849,4 +849,142 @@ export class LedgerStorage extends Storages
...
+            `SELECT
+                T.block_height as height,
+                T.tx_hash,
+                T.type,
+                T.unlock_height
+            FROM
+                (
+                    SELECT * FROM transactions WHERE tx_hash = ?
+                ) T
+                INNER JOIN blocks B ON (B.height = T.block_height);`;
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
            FROM 
                blocks B INNER JOIN transactions T
                ON (B.height = T.block_height)
            WHERE T.tx_hash= ? ;`;
```

Comment (MichaelKim20) [reply src/modules/storage/LedgerStorage.ts]: Why?

**Review by TrustHenry — APPROVED** [review]
LGTM

**Pull request merged**